File: packages/nodejs/packages/bridge/src/sinks/http.ts
import type { DeviceEnvelope } from "../envelope.js";

export class HttpSink {
  constructor(
    private readonly url: string,
    private readonly fetchFn: typeof fetch = fetch,
  ) {}

  async emit(envelope: DeviceEnvelope): Promise<void> {
    const body = JSON.stringify(envelope);
    const res = await this.fetchFn(this.url, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body,
    });
    if (!res.ok) throw new Error(`http sink ${res.status}`);
  }

  async close(): Promise<void> {}
}

File: packages/python/btw_bridge/src/btw_bridge/sinks/http.py
from __future__ import annotations

import httpx

from btw_bridge.envelope import DeviceEnvelope


class HttpSink:
    def __init__(
        self,
        url: str,
        *,
        timeout_s: float = 10.0,
        headers: dict[str, str] | None = None,
    ) -> None:
        self._url = url
        self._client = httpx.AsyncClient(timeout=timeout_s, headers=headers or {})

    async def emit(self, envelope: DeviceEnvelope) -> None:
        await self._client.post(
            self._url,
            content=envelope.model_dump_json_bytes(),
            headers={"content-type": "application/json"},
        )

    async def aclose(self) -> None:
        await self._client.aclose()
